pd/src/app.rs

```diff
@@ -160,33 +175,192 @@ impl App {
 
     #[instrument(skip(self))]
     fn begin_block(&mut self, begin: BeginBlock) -> response::BeginBlock {
-        self.pending_block = Some(PendingBlock::new(self.note_commitment_tree.clone()));
+        self.pending_block = Arc::new(Mutex::new(Some(PendingBlock::new(
+            self.note_commitment_tree.clone(),
+        ))));
         response::BeginBlock::default()
     }
 
+    /// Perform checks before adding a transaction into the mempool via `CheckTx`.
+    ///
+    /// In the transaction validation performed before adding a transaction into the
+    /// mempool, we check that:
+    ///
+    /// * All binding and auth sigs signatures verify (stateless),
+    /// * All proofs verify (stateless and stateful),
+    /// * The transaction does not reveal nullifiers already revealed in another transaction
+    /// in the mempool or in the database,
```

Comment:
The checks we want to perform at the `CheckTx` could be very minimal since it's up to us, but for now we're performing the full validation to reduce the invalid transactions that make their way into blocks

Reply:
Yeah, I think until we do ABCI++ we can't influence the selection of transactions by the block proposer, so we need to rely on `CheckTx` doing full verification.